### ISSUE TITLE
没有预测脚本吗，这个训练完之后，怎么进行推理呢？

### ISSUE THREAD

**Issue by lzh1998-jansen** [131250208/JointExtraction#82]

**Comment by 131250208** [comment]
readme里的Evaluation部分，有其他问题再看看其他closed或者已经回答过的issues。代码不难，自己看一遍也差不多懂了

**Comment by lzh1998-jansen** [comment]
> 

我尝试 直接用一段文本作为输入，例如，input=‘today is bad’，pred=model（input）,构建这样的推理代码，但是加载数据时除了问题，

**Comment by lzh1998-lzh** [comment]
> readme里的Evaluation部分，有其他问题再看看其他closed或者已经回答过的issues。代码不难，自己看一遍也差不多懂了

 rel2id_path='../data4bert/nyt_star/rel2id.json'
    model=NlpModel(rel2id_path=rel2id_path,device=device,model_state_path='/mnt/jansen/NLP/TPlinker-joint-extraction-master/tplinker/wandb/run-20221206_113841-34rp0jht/files/model_state_dict_0.pt')
    model=model.loadwts()     
    print(model)
    tokenizer = BertTokenizerFast.from_pretrained(config["bert_path"], add_special_tokens = False, do_lower_case = False)
    rel2id = json.load(open(rel2id_path, "r", encoding = "utf-8"))
    print(rel2id)
    handshaking_tagger = HandshakingTaggingScheme(rel2id = rel2id, max_seq_len = 16)
    data_maker = DataMaker4Bert(tokenizer, handshaking_tagger)
    get_tok2char_span_map = lambda text: tokenizer.encode_plus(text, return_offsets_mapping = True, add_special_tokens = False)["offset_mapping"]

    indexed_test_data=[{'text':"beijing is the captial of china"}]
    indexed_test_data = data_maker.get_indexed_data(indexed_test_data, 16, data_type = "test") 
    print(indexed_test_data)

    test_dataloader = DataLoader(MyDataset(indexed_test_data), 
                              batch_size = 1, 
                              shuffle = False, 
                              num_workers = 6,
                              drop_last = False,
                              collate_fn = lambda data_batch: data_maker.generate_batch(data_batch, data_type = "test"),
                             )
    pred_sample_list = []  
    for batch_data in tqdm(test_dataloader,desc='predicting'):
        sample_list, batch_input_ids, \
            batch_attention_mask, batch_token_type_ids, \
            tok2char_span_list, _, _, _ = batch_data
        batch_input_ids, \
            batch_attention_mask, \
            batch_token_type_ids = (batch_input_ids.to(device), 
                                      batch_attention_mask.to(device), 
                                      batch_token_type_ids.to(device))
        print(sample_list)
        with torch.no_grad():
            batch_ent_shaking_outputs, \
                batch_head_rel_shaking_outputs, \
                batch_tail_rel_shaking_outputs = model(batch_input_ids, 
                                                          batch_attention_mask, 
                                                          batch_token_type_ids, 
                                                         )
        batch_ent_shaking_tag, \
        batch_head_rel_shaking_tag, \
        batch_tail_rel_shaking_tag = torch.argmax(batch_ent_shaking_outputs, dim = -1), \
                                     torch.argmax(batch_head_rel_shaking_outputs, dim = -1), \
                                     torch.argmax(batch_tail_rel_shaking_outputs, dim = -1)
                                  
        for ind in range(len(sample_list)):
            gold_sample = sample_list[ind]
            text = gold_sample["text"]
            #text_id = gold_sample["id"]
            tok2char_span = tok2char_span_list[ind]
            ent_shaking_tag, \
            head_rel_shaking_tag, \
            tail_rel_shaking_tag = batch_ent_shaking_tag[ind], \
                                    batch_head_rel_shaking_tag[ind], \
                                    batch_tail_rel_shaking_tag[ind]
            print(ent_shaking_tag)
            print(head_rel_shaking_tag)
            print(tail_rel_shaking_tag)
            
            tok_offset, char_offset = 0, 0
            # if split_test_data:
            #     tok_offset, char_offset = gold_sample["tok_offset"], gold_sample["char_offset"]
            rel_list = handshaking_tagger.decode_rel_fr_shaking_tag(text, 
                                                                    ent_shaking_tag, 
                                                                    head_rel_shaking_tag, 
                                                                    tail_rel_shaking_tag, 
                                                                    tok2char_span, 
                                                                    tok_offset = tok_offset, char_offset = char_offset)
            pred_sample_list.append({
                "text": text,
                # "id": text_id,
                "relation_list": rel_list,
            })
            print(pred_sample_list)

**Comment by lzh1998-lzh** [comment]
rel2id_path='../data4bert/nyt_star/rel2id.json'
    model=NlpModel(rel2id_path=rel2id_path,device=device,model_state_path='/mnt/jansen/NLP/TPlinker-joint-extraction-master/tplinker/wandb/run-20221206_113841-34rp0jht/files/model_state_dict_0.pt')
    model=model.loadwts()     
    print(model)
    tokenizer = BertTokenizerFast.from_pretrained(config["bert_path"], add_special_tokens = False, do_lower_case = False)
    rel2id = json.load(open(rel2id_path, "r", encoding = "utf-8"))
    print(rel2id)
    handshaking_tagger = HandshakingTaggingScheme(rel2id = rel2id, max_seq_len = 16)
    data_maker = DataMaker4Bert(tokenizer, handshaking_tagger)
    get_tok2char_span_map = lambda text: tokenizer.encode_plus(text, return_offsets_mapping = True, add_special_tokens = False)["offset_mapping"]

    indexed_test_data=[{'text':"beijing is the captial of china"}]
    indexed_test_data = data_maker.get_indexed_data(indexed_test_data, 16, data_type = "test") 
    print(indexed_test_data)

    test_dataloader = DataLoader(MyDataset(indexed_test_data), 
                              batch_size = 1, 
                              shuffle = False, 
                              num_workers = 6,
                              drop_last = False,
                              collate_fn = lambda data_batch: data_maker.generate_batch(data_batch, data_type = "test"),
                             )
    pred_sample_list = []  
    for batch_data in tqdm(test_dataloader,desc='predicting'):
        sample_list, batch_input_ids, \
            batch_attention_mask, batch_token_type_ids, \
            tok2char_span_list, _, _, _ = batch_data
        batch_input_ids, \
            batch_attention_mask, \
            batch_token_type_ids = (batch_input_ids.to(device), 
                                      batch_attention_mask.to(device), 
                                      batch_token_type_ids.to(device))
        print(sample_list)
        with torch.no_grad():
            batch_ent_shaking_outputs, \
                batch_head_rel_shaking_outputs, \
                batch_tail_rel_shaking_outputs = model(batch_input_ids, 
                                                          batch_attention_mask, 
                                                          batch_token_type_ids, 
                                                         )
        batch_ent_shaking_tag, \
        batch_head_rel_shaking_tag, \
        batch_tail_rel_shaking_tag = torch.argmax(batch_ent_shaking_outputs, dim = -1), \
                                     torch.argmax(batch_head_rel_shaking_outputs, dim = -1), \
                                     torch.argmax(batch_tail_rel_shaking_outputs, dim = -1)
                                  
        for ind in range(len(sample_list)):
            gold_sample = sample_list[ind]
            text = gold_sample["text"]
            #text_id = gold_sample["id"]
            tok2char_span = tok2char_span_list[ind]
            ent_shaking_tag, \
            head_rel_shaking_tag, \
            tail_rel_shaking_tag = batch_ent_shaking_tag[ind], \
                                    batch_head_rel_shaking_tag[ind], \
                                    batch_tail_rel_shaking_tag[ind]
            print(ent_shaking_tag)
            print(head_rel_shaking_tag)
            print(tail_rel_shaking_tag)
            
            tok_offset, char_offset = 0, 0
            # if split_test_data:
            #     tok_offset, char_offset = gold_sample["tok_offset"], gold_sample["char_offset"]
            rel_list = handshaking_tagger.decode_rel_fr_shaking_tag(text, 
                                                                    ent_shaking_tag, 
                                                                    head_rel_shaking_tag, 
                                                                    tail_rel_shaking_tag, 
                                                                    tok2char_span, 
                                                                    tok_offset = tok_offset, char_offset = char_offset)
            pred_sample_list.append({
                "text": text,
                # "id": text_id,
                "relation_list": rel_list,
            })
            print(pred_sample_list)

**Comment by lzh1998-jansen** [comment]
> readme里的Evaluation部分，有其他问题再看看其他closed或者已经回答过的issues。代码不难，自己看一遍也差不多懂了

上面是我根据代码组合了一个预测的功能，请您看看逻辑上有问题吗，我现在加载一个训练了一次的代码，输出的 relation_list 是 空，如果没有什么问题，我就多训练几次，训练一次要好几个小时
